ledger/eval/applications.go

```diff
@@ -319,38 +343,35 @@ func (cs *roundCowState) Perform(gi int, ep *logic.EvalParams) error {
 
 	switch txn.Txn.Type {
 	case protocol.PaymentTx:
-		err = apply.Payment(txn.Txn.PaymentTxnFields, txn.Txn.Header, cs, *ep.Specials, &txn.ApplyData)
+		err = apply.Payment(txn.Txn.PaymentTxnFields, txn.Txn.Header, cowForTxn, *ep.Specials, &txn.ApplyData)
 
 	case protocol.KeyRegistrationTx:
-		err = apply.Keyreg(txn.Txn.KeyregTxnFields, txn.Txn.Header, cs, *ep.Specials, &txn.ApplyData,
-			cs.Round())
+		err = apply.Keyreg(txn.Txn.KeyregTxnFields, txn.Txn.Header, cowForTxn, *ep.Specials, &txn.ApplyData,
+			cowForTxn.Round())
 
 	case protocol.AssetConfigTx:
-		err = apply.AssetConfig(txn.Txn.AssetConfigTxnFields, txn.Txn.Header, cs, *ep.Specials, &txn.ApplyData,
-			cs.Counter())
+		err = apply.AssetConfig(txn.Txn.AssetConfigTxnFields, txn.Txn.Header, cowForTxn, *ep.Specials, &txn.ApplyData,
+			cowForTxn.Counter())
 
 	case protocol.AssetTransferTx:
-		err = apply.AssetTransfer(txn.Txn.AssetTransferTxnFields, txn.Txn.Header, cs, *ep.Specials, &txn.ApplyData)
+		err = apply.AssetTransfer(txn.Txn.AssetTransferTxnFields, txn.Txn.Header, cowForTxn, *ep.Specials, &txn.ApplyData)
 
 	case protocol.AssetFreezeTx:
-		err = apply.AssetFreeze(txn.Txn.AssetFreezeTxnFields, txn.Txn.Header, cs, *ep.Specials, &txn.ApplyData)
+		err = apply.AssetFreeze(txn.Txn.AssetFreezeTxnFields, txn.Txn.Header, cowForTxn, *ep.Specials, &txn.ApplyData)
 
 	case protocol.ApplicationCallTx:
-		err = apply.ApplicationCall(txn.Txn.ApplicationCallTxnFields, txn.Txn.Header, cs, &txn.ApplyData,
-			gi, ep, cs.Counter())
+		err = apply.ApplicationCall(txn.Txn.ApplicationCallTxnFields, txn.Txn.Header, cowForTxn, &txn.ApplyData,
+			gi, ep, cowForTxn.Counter())
 
 	default:
 		err = fmt.Errorf("%s tx in AVM", txn.Txn.Type)
 	}
-	if err != nil {
-		return err
-	}
 
 	// We don't check min balances during in app txns.
 
 	// func (eval *BlockEvaluator) checkMinBalance will take care of it when the
 	// top-level txn concludes, because cow will return all changed accounts in
 	// modifiedAccounts().
 
-	return nil
+	return
```

Comment:
I would prefer that we leave in the explicit error returns. I think naked returns are always confusing in longer functions.